File: tests/fixtures/xpath_injection/php/safe_xpath_query.php
<?php
// Safe: $_GET['user'] routed through the project-local `escape_xpath` helper
// before concatenation.  The sanitizer clears XPATH_INJECTION so the sink
// does not fire.
function escape_xpath($raw) {
    return str_replace(["'", "\""], ["&apos;", "&quot;"], $raw);
}

$xml = simplexml_load_file("users.xml");
$user = $_GET['user'];
$safe = escape_xpath($user);
$expr = "//user[name='" . $safe . "']";
$nodes = $xml->xpath($expr);

File: tests/fixtures/xpath_injection/ruby/safe_xpath_query.rb
# Safe: user-supplied substring routed through the project-local
# `escape_xpath` helper before interpolation.  The sanitizer clears
# XPATH_INJECTION so the sink does not fire.
require 'nokogiri'

def escape_xpath(raw)
  raw.gsub("'", '&apos;').gsub('"', '&quot;')
end

def lookup(params)
  doc = Nokogiri::XML(File.read("users.xml"))
  user = params[:user]
  safe = escape_xpath(user)
  expr = "//user[name='#{safe}']"
  doc.xpath(expr)
end
